### ISSUE TITLE
kopf framework stucked if cannot reach initially k8s API

### ISSUE THREAD

**Issue by brutus333** [zalando-incubator/kopf#248]
## Expected Behavior

If kopf can't reach k8s API at start it should either exit with non-zero exit code or retry after a fix/variable interval.

## Actual Behavior

kopf is stalled so there is no processing of cluster events even after k8s API recovery.

```
[2019-11-19 16:44:40,166] kopf.engines.peering [ERROR   ] Couldn't remove self from the peering. Ignoring.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/kopf/engines/peering.py", line 261, in peers_keepalive
    await asyncio.shield(ourselves.disappear())
  File "/usr/local/lib/python3.7/site-packages/kopf/engines/peering.py", line 145, in disappear
    await apply_peers([self], name=self.name, namespace=self.namespace, legacy=self.legacy)
  File "/usr/local/lib/python3.7/site-packages/kopf/engines/peering.py", line 190, in apply_peers
    await patching.patch_obj(resource=resource, namespace=namespace, name=name, patch=patch)
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/patching.py", line 34, in patch_obj
    cls = classes._make_cls(resource=resource)
  File "/usr/local/lib/python3.7/site-packages/kopf/clients/classes.py", line 10, in _make_cls
    api_resources = api.resource_list(resource.api_version)['resources']
  File "/usr/local/lib/python3.7/site-packages/pykube/http.py", line 204, in resource_list
    r.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 504 Server Error: Gateway Timeout for url: https://100.20.0.1:443/apis/zalando.org/v1/
```

## Steps to Reproduce the Problem

  It's quite hard to reproduce since it was only occurring on very busy or undersized k8s cluster.

## Specifications

  - Platform: linux
  - Kubernetes version: *(use `kubectl version`)* 1.12.4
  - Python version: *(use `python --version`)* 3.7.4
  - Python packages installed: *(use `pip freeze --all`)*
aiohttp==3.6.1
aiojobs==0.2.2
async-timeout==3.0.1
attrs==19.2.0
cachetools==3.1.1
certifi==2019.9.11
chardet==3.0.4
Click==7.0
google-auth==1.6.3
idna==2.8
iso8601==0.1.12
jsonpatch==1.24
jsonpointer==2.0
kopf==0.21
kubernetes==10.0.0
multidict==4.5.2
oauthlib==3.1.0
pip==19.2.3
pyasn1==0.4.7
pyasn1-modules==0.2.6
pykube-ng==19.9.2
python-dateutil==2.8.0
PyYAML==5.1.2
requests==2.22.0
requests-oauthlib==1.2.0
rsa==4.0
setuptools==41.2.0
six==1.12.0
urllib3==1.25.6
websocket-client==0.56.0
wheel==0.33.6
yarl==1.3.0

```
```

**Comment by nolar** [comment]
Can your please try it with [`kopf==0.23`](https://github.com/zalando-incubator/kopf/releases/tag/0.23)? There was an issue with freezing at exiting and on errors for ~30-90 seconds (#142). In 0.23, the I/O is fully reworked to asyncio+aiohttp, there are no blocking synchronous `socket.read()` threads anymore, so it should fail normally.

Retries are also possible to implement now in one central point (there are retries for HTTP 401 "Unauthorised"). But I have doubts this is a good idea per se (though discussable).

**Comment by brutus333** [comment]
Sure, I was already trying, but I probably need to understand more about the changes in 0.23 since the framework is failing initially with `ERROR    kopf.reactor.running:running.py:423 Root task 'watcher of nodes' is failed: ClientResponseError(RequestInfo(url=URL('https://127.0.0.1:53687/api/v1/nam
espaces/default/nodes'), method='GET', headers=<CIMultiDictProxy('Host': '127.0.0.1:53687', 'User-Agent': 'kopf/unknown', 'Accept': '*/*', 'Accept-Encoding': 'g
zip, deflate')>, real_url=URL('https://127.0.0.1:53687/api/v1/namespaces/default/nodes')), (), status=404, message='Not Found', headers=<CIMultiDictProxy('Conte
nt-Type': 'application/json', 'Date': 'Wed, 20 Nov 2019 14:22:50 GMT', 'Content-Length': '174')>)
FAILED                             `. 
Thank you

**Comment by brutus333** [comment]
Regarding the error I posted earlier, it is occurring when the framework is started with `--namespace` argument. Is it expected?

**Comment by nolar** [comment]
`/api/v1/namespaces/default/nodes` — definitely not expected. The nodes are not namespaced. It should be a cluster query.

Seems, the cluster resources was broken when switched off `pykube-ng` — it was doing the pre-listing of the resources and parsed their scope.

Let's see if I can quickly fix this... Created #249 for this.

**Comment by nolar** [comment]
@brutus333 The 404 error should be fixed in [0.23.1](https://github.com/zalando-incubator/kopf/releases/tag/0.23.1).

**Comment by brutus333** [comment]
Thank you for your very quick response and fix. I am highly appreciating your code quality.

**Comment by nolar** [comment]
@brutus333 Can you please re-check the original issue with `kopf>=0.23.2` (stalled operator on AP error)? Does it re-appear? Or maybe with a different exception (it should be `aiohttp` error now)?

**Comment by brutus333** [comment]
Hello!
I am trying to reproduce the issue. Not sure however if I can succeed. I had to add some extra latency to API responses... Maybe I can use for this toxiproxy.
I'll let you know once I have a result.
Thank you.
Virgil

**Comment by brutus333** [comment]
I managed to reproduce partially the conditions and it seems that 0.23.2 is behaving better.
I used a kind cluster and I issued a kill -STOP api process (restarted later by systemd). In kopf 0.21 and 0.22 the operator exited only when remote reset the connection while in kopf 0.23.2 the operator exited after 60 seconds with a clear message: `SSL handshake is taking longer than 60.0 seconds: aborting the connection`.

I uploaded the logs for the three versions:

[scratch_19.txt](https://github.com/zalando-incubator/kopf/files/3940224/scratch_19.txt)

In usual k8s case (with several k8s masters and calico networking) since kopf was connecting to kube api server via a service most probably the connection reset was not issued since a new pod was created (and no other pod took over old pod IP).
